### PR TITLE
Edit a CustomResource should result in a patch

### PR DESCRIPTION
**Pull request by andreaTP**
## Description
Trying to use the `edit` of a CR I found a couple of issues:
 - it throws `ClassCastException` if there are external json modules registered e.g.: https://github.com/andreaTP/fabric8-edit-as-patch/blob/master/src/main/scala/Main.scala#L59
 - edit perform a `PUT` while it's supposed to be a `PATCH` as per: https://github.com/fabric8io/kubernetes-client/blob/8d0957cac72cc27083a53faa84a7e80a7470b8e1/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/HasMetadataOperation.java#L37-L39

this PR address both issues.

## Type of change
 - [X] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [X] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [ ] I Added [CHANGELOG](../CHANGELOG.md) entry regarding this change
 - [X] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [X] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

I tested the final result with this demo project:
https://github.com/andreaTP/fabric8-edit-as-patch
run with: `sbt run`

I don't know how/where to add more exhaustive tests, I'm looking forward to guidance here 🙏

### CODE REVIEW

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/RawCustomResourceOperationsImpl.java
Comment (andreaTP) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/RawCustomResourceOperationsImpl.java:350]: This is already implemented here:
https://github.com/fabric8io/kubernetes-client/blob/6e439a876f3b7973c744aa35bece70aca4700f39/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/RawCustomResourceOperationsImpl.java#L882

**Comment by rohanKanojia** [comment]
A test seems to be failing:
```
[INFO] Results:
[INFO] 
Error:  Errors: 
Error:    CustomResourceTest.testEdit:197 » KubernetesClient Failure executing: PATCH at...
[INFO] 
Error:  Tests run: 670, Failures: 0, Errors: 1, Skipped: 4
[INFO] 
[INFO] ------------------------------------------------------------------------
```

**Comment by andreaTP** [comment]
@rohanKanojia I should have fixed the failing test, I'm having troubles with the build process, it often fails ... any hint on how to get it more reliable?

**Comment by rohanKanojia** [comment]
You mean running tests via ide? I agree it's painful. Could you please check if this helps https://github.com/fabric8io/kubernetes-client/pull/2676#issuecomment-747947856

**Comment by andreaTP** [comment]
yesterday I was working in the IDE, today the IDE stopped to be able to compile the project and I have had to test every time with a `mvn clean install` command (and crazy long feedback loop cycle as you can imagine ...).
Re-importing failed at least 3 times ...

**Comment by rohanKanojia** [comment]
You don't have to compile everything every time. Since you've made changes to `kubernetes-client` only, you should be able to check `kubernetes-client` and `kubernetes-tests` build okay.

However, you might need to build a project the first time. I use `-DskipTests` which compiles project in 10 minutes

**Comment by andreaTP** [comment]
I think that the CI failure is unrelated now

**Comment by rohanKanojia** [comment]
Don't worry, I've restarted the workflow

**Review by rohanKanojia — APPROVED** [review]

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/RawCustomResourceOperationsImpl.java
Comment (manusa) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/RawCustomResourceOperationsImpl.java:99]: Why can't we reuse the global ObjectMapper singleton instance?

Comment (andreaTP) [reply kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/RawCustomResourceOperationsImpl.java:99]: You can test from my example:
https://github.com/andreaTP/fabric8-edit-as-patch/blob/8035f945a3c0be2db239c92632dc91bb4dda6d9e/src/main/scala/Main.scala#L59

After registering `DefaultScalaModule` the global ObjectMapper will deserialize as scala Maps resulting in class cast exception e.g. here:
https://github.com/fabric8io/kubernetes-client/blob/e60c6348bb3730b2c909aee4f71cbf5db86cc3f7/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/RawCustomResourceOperationsImpl.java#L870

Comment (manusa) [reply kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/RawCustomResourceOperationsImpl.java:99]: OK, that's very specific for your use case, and I'm sure you are having issues in other places around the code-base.

We should really need to find a neater way to address the problem.

Especially because if we start creating ObjectMapper instances across the different service implementations we'll have more problems in the future.

Comment (andreaTP) [reply kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/RawCustomResourceOperationsImpl.java:99]: Please just propose me a solution, I will go-ahead 🙂 
Here we _RELY_ on having a clean `ObjectMapper`, but, as said I'm open to suggestions!

Just to mention, no, fortunately, I have no other Serialization issues:
 - [Cloudflow Operator](https://github.com/lightbend/cloudflow/tree/master/tools/cloudflow-operator/src/main/scala/cloudflow/operator) handling 3 different CR, one CRD, ConfigMaps, Secrets, Volumes etc. etc.
 - [Cloudflow CLI](https://github.com/lightbend/cloudflow/blob/master/tools/cloudflow-cli/src/main/scala/akka/cli/cloudflow/kubeclient/KubeClientFabric8.scala) reading and writing again CRs, Secrets and more

Comment (andreaTP) [reply kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/RawCustomResourceOperationsImpl.java:99]: @manusa any suggestion to unblock this PR?
should we use the `PatchUtils` `ObjectMapper`? https://github.com/fabric8io/kubernetes-client/blob/3ed4809e50f2e71a362487318331ae66bf644931/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/internal/PatchUtils.java#L40

I still think that a fresh one is more appropriate here until we have a more "structural" solution.
WDYT?

**Comment by andreaTP** [comment]
rebased on the latest master, anything else I can do to get this moving?

**Comment by rohanKanojia** [comment]
@andreaTP : Thanks for updating as per review comments. I think we can get this merged once Marc approves it.

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/RawCustomResourceOperationsImpl.java
Comment (manusa) [reply kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/RawCustomResourceOperationsImpl.java:99]: Yes, that seems very reasonable.

**Review by manusa — APPROVED** [review]
LGTM, thx!

**Pull request merged**
